### ISSUE TITLE
Automated build don't get run automatically

### ISSUE THREAD

**Issue by mathias-luedtke** [ros-industrial/docker#14]
The Docker images are way older than the Dockerfiles.
It looks like to hook is missing that notifies Docker Hub on pushes.

PS: I will trigger the CI image build manually.

**Comment by AustinDeric** [comment]
good catch.  I intended to for dockerhub to trigger new builds whenever this repo was pushed.  

![image](https://user-images.githubusercontent.com/5349043/30710542-8c00ab06-9ecb-11e7-91e8-02ffe8f3d3a3.png)


I tested this functionality, i am not sure why its not working now.

**Comment by 130s** [comment]
Can I offer to take a look at this? Just more eyes and hands, it's not that I'm an expert for Docker Hub though.

I see at least https://hub.docker.com/r/rosindustrial/ci/builds/ seems to be triggered to build very frequently (unless someone's been manually doing that).

My hub id: d130s

**Comment by mathias-luedtke** [comment]
> Can I offer to take a look at this? Just more eyes and hands, it's not that I'm an expert for Docker Hub though.

Sure, you're more than welcome!
What's your Docker username?

> I see at least https://hub.docker.com/r/rosindustrial/ci/builds/ seems to be triggered to build very frequently (unless someone's been manually doing that).

At some point I configured the `ci` build to depend on the [ros](https://hub.docker.com/_/ros/) images, so it looks like it's working now.

**Comment by 130s** [comment]
Thanks, I included my id by updating my previous post.

**Comment by 130s** [comment]
I noticed today that I have an access to settings on https://hub.docker.com/r/rosindustrial/, so thanks.

> At some point I configured the `ci` build to depend on the `ros` images, so it looks like it's working now.

I manually triggered build on `core` (which was built 8 months ago the last time) but errored. After that I added `ros` to `core` too. Let's see.

**Comment by 130s** [comment]
I saw `core` built a day ago, so I think the build is automated unless someone manually triggered.

It just needed to add `ros` in the linked repos as in this image:
![screenshot from 2018-01-17 09-47-13](https://user-images.githubusercontent.com/1840401/35058039-8fa08d6a-fb6b-11e7-8114-5b1d83b4eb33.png)

**Comment by mathias-luedtke** [comment]
> It just needed to add ros in the linked repos as in this image:

The `ci` images have been built as well.
We should check if something changes after the next merge of a Dockerfile.

**Comment by AustinDeric** [comment]
We just merged a Dockerfile in the ci repo and the builds didn't automatically trigger, i am not sure why at this point: https://hub.docker.com/r/rosindustrial/ci/builds/

#18

**Comment by mathias-luedtke** [comment]
@AustinDeric: please review the repostory settings and check the integration/service.

For now you can add Docker service (if possible), not sure how it works with apps.

**Comment by mathias-luedtke** [comment]
https://docs.docker.com/docker-hub/github/#github-organizations

**Comment by AustinDeric** [comment]
- Thanks @ipa-mdl that may be the issue, but my permission level doesn't allow me to access any of that.

 - @Levi-Armstrong or @JeremyZoss  can you give me the settings tab back for this repo?

 - Worst case scenario, it looks like we can manually add webhooks for each build:

>  GitHub Service hook allows GitHub to notify the Docker Hub when something has been committed to a given git repository.
> 
> When you create an Automated Build from a GitHub user that has full “Public and Private” linking, a Service Hook should get automatically added to your GitHub repository.
> 
> If your GitHub account link to the Docker Hub is “Limited Access”, then you need to add the Service Hook manually.
> 
> To add, confirm, or modify the service hook, log in to GitHub, then navigate to the repository, click “Settings” (the gear), then select “Webhooks & Services”. You must have Administrator privileges on the repository to view or modify this setting.
> 
> The image below shows the “Docker” Service Hook.
> 
> ![image](https://user-images.githubusercontent.com/5349043/40730408-298b2e8a-63f4-11e8-8133-61e9c8a1b461.png)
>

**Comment by gavanderhoorn** [comment]
That is not on the repository level, but at the organization level.

I just checked and there are no third-party applications listed on that panel for `ros-industrial`.

But there are also no restrictions configured. Not sure how that influences things.

**Comment by gavanderhoorn** [comment]
You are still a member of the `docker-admins` team btw, so you should have access to [ros-industrial/docker/settings](https://github.com/ros-industrial/docker/settings).

**Comment by AustinDeric** [comment]
- Nope no access to https://github.com/ros-industrial/docker/settings 

 - I am also not able to link the ros-industrial organization's repo to auto-build:

![image](https://user-images.githubusercontent.com/5349043/40730638-bf377308-63f4-11e8-96bd-9d9b77b2c100.png)

**Comment by gavanderhoorn** [comment]
Seems `docker admins` only had `Write` access. I just changed that (back?) to `Admin`.

This could have happened when we were cleaning up access privileges: as you are no longer an SwRI employee this got reset.

**Comment by AustinDeric** [comment]
Settings tab is there now! Thanks!

However i still cannot link the ros-industrial organization's repo as shown in my post above.

**Comment by gavanderhoorn** [comment]
You're only admin on this specific repository, and are not an owner (of the org). I don't know whether that would be required.

**Comment by gavanderhoorn** [comment]
Looking at the third party applications settings panel, I see this:
> As members request access for specific applications, those requests will be listed here for your approval. You can start by browsing [your own authorized applications](https://github.com/settings/applications#authorized).

**Comment by AustinDeric** [comment]
In the meantime, I am going to setup a `test` branch and `test` automated build so we can test this functionality frequently without adulterating the master branch.

**Comment by AustinDeric** [comment]
I don't believe i am able to request access.  ros-industrial is not listed in link you posted (https://github.com/settings/applications#authorized).  Also, in the last image i posted above, ros-industrial is not listed.

**Comment by gavanderhoorn** [comment]
@AustinDeric wrote:
> ros-industrial is not listed in link you posted (https://github.com/settings/applications#authorized). 

I don't believe it should be. Docker Hub should be listed there, if I understand things correctly.

**Comment by JeremyZoss** [comment]
From your screenshot, it looks like you have the desired access on `ros-industrial-consortium` and `ros-industrial-release` organizations.  It might be worth comparing the settings for those organizations with `ros-industrial` to see if anything jumps out.

I only have access to `ros-industrial-release` and `ros-industrial`.  The only relevant difference I see in the Settings pages is under "Third-Party Access".  `ros-industrial` shows "Access Restricted", and `ros-industrial-release` shows "No restrictions".

[This](https://help.github.com/articles/requesting-organization-approval-for-oauth-apps/) page suggests that one process might be to add Docker Hub to your personal GitHub account, and then Request Access for the app to the ros-industrial organization.  Then it seems like it'll show up in the proper ros-industrial settings page for Gijs or I to approve.  Just a shot in the dark, since I don't have much experience with Docker or GitHub apps...

**Comment by gavanderhoorn** [comment]
@JeremyZoss wrote:
> I only have access to `ros-industrial-release` and `ros-industrial`.  The only relevant difference I see in the Settings pages is under "Third-Party Access".  `ros-industrial` shows "Access Restricted", and `ros-industrial-release` shows "No restrictions".

yeah, I set the restrictions. Not to be difficult, but to see whether something that @AustinDeric did would make the auth request show up.

**Comment by AustinDeric** [comment]
> This page suggests that one process might be to add Docker Hub to your personal GitHub account, and then Request Access for the app to the ros-industrial organization. Then it seems like it'll show up in the proper ros-industrial settings page for Gijs or I to approve. Just a shot in the dark, since I don't have much experience with Docker or GitHub apps...

That did the trick, apparently you had to click on the app and from there you can request access.  Thanks guys!

**Comment by gavanderhoorn** [comment]
Now I see something appearing.

Just curious: before I click `Grant access`, can you see what Docker can currently do without granting access to our 'private data'?

**Comment by AustinDeric** [comment]
> can you see what Docker can currently do without granting access to our 'private data'?

Please clarify if i misunderstand your Question.  I interpret this as, how can we accomplish automated builds without authorizing this Oauth app.  If so, the answer is to manually create webhooks with which github notifies docker hub via a POST message.  https://docs.docker.com/docker-hub/github/#github-private-submodules (scroll down to "GitHub service hooks").  Its not as elegant as the integration I am proposing which is identical to travis.

**Comment by gavanderhoorn** [comment]
I'm not suggesting that at all.

I'm just curious what granting access changes.

The message I received stated:

> Until it is approved, this application will have no access to private resources and will have read-only access to public resources belonging to your organization.

So there should already be *some* access. It's just not clear how much.

**Comment by gavanderhoorn** [comment]
I'm perfectly ok with granting this, but from the documentation (both on Github as well as on the Docker site), it looks like manually adding the webhook would enable Travis-like integration just as well.

> When enabled this will let the Docker Hub know when you have made changes to
your repository. The Docker Hub will then pull down your repository and build
your Dockerfile to create a Docker repository and push it onto the Docker Hub
so that it is available for others to download. When you commit changes to your
git Repo the Docker Hub will keep the Docker Repository up to date.

What I can't find though is whether 'cron' builds on Docker are supported that way.

You should have access to this as well btw: https://github.com/ros-industrial/docker/settings/installations.

---

Edit: from the Docker hub docs you linked earlier:

> If you selected to link your GitHub account with only a “Limited Access” link, then after creating your automated build, you need to either manually trigger a Docker Hub build using the “Start a Build” button, **or** add the GitHub webhook manually

So adding the webhook should work?

**Comment by AustinDeric** [comment]
![image](https://user-images.githubusercontent.com/5349043/40739570-2f27ecd0-640c-11e8-9d8b-03c30ba81005.png)

https://help.github.com/articles/about-oauth-app-access-restrictions/

**Comment by gavanderhoorn** [comment]
Btw, according to [this](https://success.docker.com/article/why-cant-i-create-automated-builds-for-my-organization), you'll probably need to set your membership of `ros-industrial` to `public`.

**Comment by gavanderhoorn** [comment]
Ok, so I get the impression from the Docker docs -- which are far from clear tbh -- that it's impossible to configure automated builds for repositories -- even public ones -- on organizations without granting access to 'everything'.

I'm actually not too much of a fan of this, but I see the appeal of automated builds for these Docker images.

I'd like to see a `+1` from one other owner here on `ros-industrial` before I click the button. I realise Docker promises 'nothing else will be touched', and they're probably OK, but still.

@JeremyZoss, @Levi-Armstrong

**Comment by JeremyZoss** [comment]
I'm not very familiar with the various methods GitHub uses to connect with external 3rd-party apps.  But this Docker approach seems different than what's used for Travis CI, for example.

It seems to me that this elevated access is required for Docker to auto-add the "Docker Service" hook to specific repositories.  Since the only repo we're considering at this point is the docker repo (??), it wouldn't be too much effort for Austin to add this hook [manually](https://docs.docker.com/docker-hub/github/#github-service-hooks).

As another point of reference, GitHub is [deprecating](https://developer.github.com/changes/2018-04-25-github-services-deprecation/) the [Services](https://docs.docker.com/docker-hub/github/#github-service-hooks) interface used by Docker.  3rd party services that use this API will migrate to new interfaces by the end of this year.  See [Travis CI's](https://docs.travis-ci.com/user/legacy-services-to-github-apps-migration-guide/) guide to migrating over to their GitHub Apps interface.  While I don't find a similar discussion on Docker Hub's docs, it may be that they will also be moving to a different integration method later this year.

In the end, though, you have my +1 for approving Docker's access.  I'm agreed that this is not as fine-grained of access control as we'd like, but it's "probably" harmless.  We should keep an eye on Docker's GitHub integration periodically throughout this year, though, to see if we can later "downgrade" to a more finely-scoped interface.

**Comment by 130s** [comment]
On docker hub, looks like I have more admin accesses on `ci` section, but much limited on others.

CI
![screenshot from 2018-06-13 15-29-04](https://user-images.githubusercontent.com/1840401/41381955-cdd26f6e-6f1e-11e8-9c33-14e0c4e6c740.png)

Others (e.g. ros-core-nvidia)
![screenshot from 2018-06-13 15-29-19](https://user-images.githubusercontent.com/1840401/41381959-d1e5d456-6f1e-11e8-80b3-ee442b074119.png)

And as you see in ros-core-nvidia, which docker images have not been built for an year or so, the source repo doesn't seem to be registered.

**Comment by mathias-luedtke** [comment]
Docker hub does not have such fine-grained restrictions.
However, for setting up automated builds, you need certain privileges on Github.
The `ci` section has been set up by @AustinDeric initially.

**Comment by AustinDeric** [comment]
Hey guys, i still don't have access which means i can't add images to docker hub:
![image](https://user-images.githubusercontent.com/5349043/42770617-99fafc52-88ea-11e8-9231-c1438b2f5e9e.png)

**Comment by 130s** [comment]
Just FYI, I found the last build for the images under `ci` tags was month old so manually triggered some build jobs (which seem to have finished fine).

(Not sure if this is the cause, but @geoffreychiou and I have seen the CI jobs where we use the Docker image from `ci` here took 60+ mins today, at `rosdep` process. It often even returned the following error. After the Docker image re-built, we saw rosdep process went fast enough.)

```
W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty-updates/main/binary-amd64/Packages  Hash Sum mismatch
```

**Comment by mathias-luedtke** [comment]
Can somebody with proper permission please check if a webhook is registered [here](https://github.com/ros-industrial/docker/settings/hooks)

The Docker Hub settings don't show any active hooks. Perhaps we have to set it up manually.

**Comment by JeremyZoss** [comment]
I just checked, and there are no webhooks currently registered at ros-industrial/docker.

**Comment by mathias-luedtke** [comment]
Thanks @JeremyZoss!

Docker rolled out their new build system [recently](https://blog.docker.com/2018/12/the-new-docker-hub/), perhaps it makes sense to relink the organization now and try if the new system work better..

**Comment by JeremyZoss** [comment]
Looking back through this issue history, it looks like both @gavanderhoorn and I agreed (hesitantly) with approving the Docker Hub authorization in GitHub ros-industrial org.  But the settings show that the request is still pending approval.  I went ahead and approved the pending request.  See if that changes anything on the Docker Hub side.
